File: cp/LC1743-Restore-the-Array-From-Adjacent-Pairs2.py
from typing import List
from collections import defaultdict


class Solution:
    def restoreArray(self, adjacentPairs: List[List[int]]) -> List[int]:
        graph = defaultdict(list)

        for x, y in adjacentPairs:
            graph[x].append(y)
            graph[y].append(x)

        root = None
        for node in graph:
            if len(graph[node]) == 1:
                root = node
                break

        nums = []
        nums.append(root)
        curr = root
        prev = None

        while len(nums) < len(graph):
            for nb in graph[curr]:
                if nb != prev:
                    nums.append(nb)
                    prev = curr
                    curr = nb
                    break

        return nums


def main():
    n = int(input())
    # adjPairs = [[int(input()) for x in range(2)] for _ in range(n)]
    adjPairs = []
    for _ in range(n):
        row = list(map(int, input().strip().split()))
        adjPairs.append(row)
    sol = Solution()
    nums = sol.restoreArray(adjPairs)

    for num in nums:
        print(num, end=" ")
    print()


if __name__ == "__main__":
    main()

File: cp/LC1743-Restore-the-Array-From-Adjacent-Pairs2.cpp
#include <bits/stdc++.h>
using namespace std;
#define ll long long
#define mod 1000000007

class Solution
{
public:
    vector<int> restoreArray(vector<vector<int>> &adjacentPairs)
    {
        unordered_map<int, vector<int>> graph;
        for (auto &edge : adjacentPairs)
        {
            graph[edge[0]].push_back(edge[1]);
            graph[edge[1]].push_back(edge[0]);
        }

        int root = 0;
        for (auto &pr : graph)
        {
            if (pr.second.size() == 1)
            {
                root = pr.first;
                break;
            }
        }
        vector<int> nums = {root};
        int prev = INT_MAX;
        while (nums.size() < graph.size())
        {
            for (auto &nb : graph[root])
            {
                if (nb != prev)
                {
                    nums.push_back(nb);
                    prev = root;
                    root = nb;
                    break;
                }
            }
        }
        return nums;
    }
};

int main()
{
    int n;
    cin >> n;
    vector<vector<int>> adjPairs;
    for (int i = 0; i < n; i++)
    {
        int x, y;
        cin >> x >> y;
        adjPairs.push_back({x, y});
    }

    Solution sol;
    vector<int> nums = sol.restoreArray(adjPairs);
    for (auto &num : nums)
    {
        cout << num << " ";
    }
    cout << endl;

    return 0;
}